cassiopeia/tools/branch_length_estimator.py

```diff
@@ -0,0 +1,325 @@
+import abc
+import copy
+from typing import List, Tuple
+
+import cvxpy as cp
```

Comment:
Agreed!

QQ: I noticed many of the requirements have a minimum version requirement, e.g. `numpy > 1.17`, is there a reason for that? I was thinking to just add `cvxpy` without a version requirement.

Reply:
Having version requirements is helpful for a couple of reasons - 

(1) Sometimes, APIs are different between versions (e.g., networkx changes the way it calls some functionality in versions <2 and >2, I believe)
(2) Sometimes, earlier versions of packages throw warnings (this is especially relevant if, for example, a particular package relies on another package that has updated and is deprecating some feature). Requiring a minimal version (or maximal version, for that matter) may help you get rid of warnings being thrown around.

In general, it's likely that having an outdated version of a particular package will be fine and the worst thing that will happen is that a warning might be thrown. However, I think it's good for software reproducibility so that when a new user comes along they can create a very similar environment to the developer. Hope this makes sense!